File: scripts/journal_stats.py
#!/usr/bin/env python3
"""
Deterministic statistics over journal/trades.jsonl for the retrospective skill.

Pairs entry/exit records by trade_id and computes per-group win rates, avg P&L,
and expectancy. Groups by trade_type, theme_tag, sector, exit_reason, conviction
bucket, and the (catalyst_occurred × mechanism_worked) failure-mode grid.

Standard library only.

Usage:
    python3 scripts/journal_stats.py                    # all-time
    python3 scripts/journal_stats.py --window 7d        # last 7 days
    python3 scripts/journal_stats.py --window 30d       # last 30
    python3 scripts/journal_stats.py --since 2026-01-01
    python3 scripts/journal_stats.py --format=json

Exit codes:
    0 — stats produced
    1 — no closed trades in the window
    2 — fatal (file missing, bad arg)
"""

import argparse
import json
import os
import sys
from collections import defaultdict
from datetime import datetime, timedelta, timezone
from typing import Any

JOURNAL_DIR = os.environ.get("JOURNAL_DIR", "journal")
TRADES_PATH = os.path.join(JOURNAL_DIR, "trades.jsonl")

CONVICTION_BUCKETS = [(0, 4, "low"), (5, 7, "mid"), (8, 10, "high")]


def _read_jsonl(path: str) -> list[dict]:
    if not os.path.exists(path):
        return []
    rows = []
    with open(path) as f:
        for line in f:
            line = line.strip()
            if line:
                rows.append(json.loads(line))
    return rows


def parse_window(window: str | None, since: str | None) -> datetime | None:
    if since:
        return datetime.fromisoformat(since).replace(tzinfo=timezone.utc)
    if not window:
        return None
    if window.endswith("d"):
        days = int(window[:-1])
        return datetime.now(timezone.utc) - timedelta(days=days)
    if window.endswith("w"):
        weeks = int(window[:-1])
        return datetime.now(timezone.utc) - timedelta(weeks=weeks)
    raise ValueError(f"unrecognized --window format: {window}")


def pair_trades(records: list[dict]) -> list[dict]:
    """Return closed trades: entry merged with its exit record."""
    entries: dict[str, dict] = {}
    pairs = []
    for r in records:
        tid = r.get("trade_id")
        if not tid:
            continue
        if r.get("type") == "entry":
            entries[tid] = r
        elif r.get("type") == "exit":
            entry = entries.get(tid)
            if not entry:
                continue  # unmatched exit; skip
            pairs.append({"entry": entry, "exit": r, "trade_id": tid})
    return pairs


def conviction_bucket(c: int | None) -> str:
    if c is None:
        return "unknown"
    for lo, hi, label in CONVICTION_BUCKETS:
        if lo <= c <= hi:
            return label
    return "unknown"


def compute_cluster_stats(pairs: list[dict]) -> dict:
    if not pairs:
        return {"count": 0}
    pnls = [p["exit"].get("pnl_pct", 0) for p in pairs]
    wins = [p for p in pnls if p > 0]
    losses = [p for p in pnls if p <= 0]
    n = len(pairs)
    win_rate = len(wins) / n if n else 0
    avg_win = sum(wins) / len(wins) if wins else 0
    avg_loss = sum(losses) / len(losses) if losses else 0
    expectancy = win_rate * avg_win + (1 - win_rate) * avg_loss
    days_held = [p["exit"].get("days_held") for p in pairs if p["exit"].get("days_held") is not None]
    return {
        "count": n,
        "win_rate_pct": round(win_rate * 100, 1),
        "avg_pnl_pct": round(sum(pnls) / n, 2),
        "median_pnl_pct": round(sorted(pnls)[n // 2], 2),
        "avg_win_pct": round(avg_win, 2),
        "avg_loss_pct": round(avg_loss, 2),
        "expectancy_pct": round(expectancy, 2),
        "avg_days_held": round(sum(days_held) / len(days_held), 1) if days_held else None,
        "trade_ids": [p["trade_id"] for p in pairs[:10]],  # cap list to avoid bloat
    }


def group_by(pairs: list[dict], key_fn) -> dict[Any, list[dict]]:
    groups: dict[Any, list[dict]] = defaultdict(list)
    for p in pairs:
        key = key_fn(p)
        groups[key].append(p)
    return groups


def main() -> int:
    p = argparse.ArgumentParser(description="Deterministic stats over journal/trades.jsonl")
    p.add_argument("--window", help="Time window: Nd (days) or Nw (weeks). Exit date must fall in window.")
    p.add_argument("--since", help="ISO date (YYYY-MM-DD); exit on or after this date.")
    p.add_argument("--format", choices=["text", "json"], default="text")
    p.add_argument("--trades", default=TRADES_PATH)
    args = p.parse_args()

    records = _read_jsonl(args.trades)
    if not records:
        print("no journal entries", file=sys.stderr)
        return 2

    window_start = parse_window(args.window, args.since)
    pairs = pair_trades(records)

    if window_start:
        def in_window(p):
            ts = p["exit"].get("timestamp", "")
            try:
                t = datetime.fromisoformat(ts.replace("Z", "+00:00"))
                return t >= window_start
            except ValueError:
                return False
        pairs = [p for p in pairs if in_window(p)]

    if not pairs:
        print("no closed trades in window", file=sys.stderr)
        return 1

    overall = compute_cluster_stats(pairs)

    groups = {
        "by_trade_type": {k: compute_cluster_stats(v) for k, v in group_by(
            pairs, lambda p: p["entry"].get("trade_type", "unknown")
        ).items()},
        "by_theme_tag": {k or "untagged": compute_cluster_stats(v) for k, v in group_by(
            pairs, lambda p: p["entry"].get("theme_tag")
        ).items()},
        "by_sector": {k: compute_cluster_stats(v) for k, v in group_by(
            pairs, lambda p: p["entry"].get("sector") or p["exit"].get("sector") or "unknown"
        ).items()},
        "by_exit_reason": {k: compute_cluster_stats(v) for k, v in group_by(
            pairs, lambda p: p["exit"].get("exit_reason", "unknown")
        ).items()},
        "by_conviction": {k: compute_cluster_stats(v) for k, v in group_by(
            pairs, lambda p: conviction_bucket(p["entry"].get("conviction"))
        ).items()},
        "by_failure_mode": {
            f"{c}|{m}": compute_cluster_stats(v)
            for (c, m), v in group_by(
                pairs,
                lambda p: (p["exit"].get("catalyst_occurred", "unknown"),
                           p["exit"].get("mechanism_worked", "unknown"))
            ).items()
        },
    }

    payload = {
        "generated_at": datetime.now(timezone.utc).isoformat(timespec="seconds"),
        "window": args.window or args.since or "all-time",
        "n_closed": len(pairs),
        "overall": overall,
        "groups": groups,
    }

    if args.format == "json":
        print(json.dumps(payload, indent=2, ensure_ascii=False))
        return 0

    # text output
    lines = []
    lines.append(f"📊 JOURNAL STATS — window: {payload['window']}   closed trades: {payload['n_closed']}")
    lines.append("=" * 72)
    lines.append(f"Overall: win {overall['win_rate_pct']}%  avg {overall['avg_pnl_pct']}%  "
                 f"expectancy {overall['expectancy_pct']}%  avg days held {overall['avg_days_held']}")
    for group_name, rows in groups.items():
        lines.append("")
        lines.append(f"-- {group_name} --")
        for key, r in sorted(rows.items(), key=lambda kv: -kv[1]["count"]):
            lines.append(f"  {str(key):<40} n={r['count']:>3}  win {r['win_rate_pct']:>5}%  "
                         f"avg {r['avg_pnl_pct']:>6.2f}%  exp {r['expectancy_pct']:>6.2f}%")
    print("\n".join(lines))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/journal_stats.mjs
#!/usr/bin/env node
/**
 * Deterministic statistics over journal/trades.jsonl for the retrospective skill.
 *
 * Pairs entry/exit records by trade_id and computes per-group win rates, avg
 * P&L, and expectancy. Groups by trade_type, theme_tag, sector, exit_reason,
 * conviction bucket, and the (catalyst_occurred × mechanism_worked) failure-
 * mode grid.
 *
 * Node 18+ stdlib only.
 *
 * Usage:
 *     node scripts/journal_stats.mjs                  # all-time
 *     node scripts/journal_stats.mjs --window 7d      # last 7 days
 *     node scripts/journal_stats.mjs --window 30d     # last 30
 *     node scripts/journal_stats.mjs --since 2026-01-01
 *     node scripts/journal_stats.mjs --format=json
 *
 * Exit codes:
 *     0 — stats produced
 *     1 — no closed trades in the window
 *     2 — fatal (file missing, bad arg)
 */

import { openStore } from './store.mjs';

const CONVICTION_BUCKETS = [
  [0, 4, 'low'],
  [5, 7, 'mid'],
  [8, 10, 'high'],
];

function parseWindow(windowStr, since) {
  if (since) {
    const d = new Date(since);
    if (isNaN(d.getTime())) throw new Error(`bad --since: ${since}`);
    return d;
  }
  if (!windowStr) return null;
  const m = /^(\d+)([dw])$/.exec(windowStr);
  if (!m) throw new Error(`unrecognized --window format: ${windowStr}`);
  const n = parseInt(m[1], 10);
  const unit = m[2];
  const ms = unit === 'd' ? n * 86400_000 : n * 7 * 86400_000;
  return new Date(Date.now() - ms);
}

function pairTrades(records) {
  const entries = new Map();
  const pairs = [];
  for (const r of records) {
    const tid = r.trade_id;
    if (!tid) continue;
    if (r.type === 'entry') entries.set(tid, r);
    else if (r.type === 'exit') {
      const entry = entries.get(tid);
      if (!entry) continue;
      pairs.push({ entry, exit: r, trade_id: tid });
    }
  }
  return pairs;
}

function convictionBucket(c) {
  if (c == null) return 'unknown';
  for (const [lo, hi, label] of CONVICTION_BUCKETS) {
    if (c >= lo && c <= hi) return label;
  }
  return 'unknown';
}

function round(n, d) {
  const k = Math.pow(10, d);
  return Math.round(n * k) / k;
}

function computeClusterStats(pairs) {
  if (!pairs.length) return { count: 0 };
  const pnls = pairs.map(p => p.exit.pnl_pct ?? 0);
  const wins = pnls.filter(p => p > 0);
  const losses = pnls.filter(p => p <= 0);
  const n = pairs.length;
  const winRate = n ? wins.length / n : 0;
  const avgWin = wins.length ? wins.reduce((a, b) => a + b, 0) / wins.length : 0;
  const avgLoss = losses.length ? losses.reduce((a, b) => a + b, 0) / losses.length : 0;
  const expectancy = winRate * avgWin + (1 - winRate) * avgLoss;
  const daysHeld = pairs
    .map(p => p.exit.days_held)
    .filter(d => d != null);
  const sortedPnls = [...pnls].sort((a, b) => a - b);
  return {
    count: n,
    win_rate_pct: round(winRate * 100, 1),
    avg_pnl_pct: round(pnls.reduce((a, b) => a + b, 0) / n, 2),
    median_pnl_pct: round(sortedPnls[Math.floor(n / 2)], 2),
    avg_win_pct: round(avgWin, 2),
    avg_loss_pct: round(avgLoss, 2),
    expectancy_pct: round(expectancy, 2),
    avg_days_held: daysHeld.length
      ? round(daysHeld.reduce((a, b) => a + b, 0) / daysHeld.length, 1)
      : null,
    trade_ids: pairs.slice(0, 10).map(p => p.trade_id),
  };
}

function groupBy(pairs, keyFn) {
  const groups = new Map();
  for (const p of pairs) {
    const k = keyFn(p);
    if (!groups.has(k)) groups.set(k, []);
    groups.get(k).push(p);
  }
  return groups;
}

function mapGroups(pairs, keyFn) {
  const out = {};
  for (const [k, v] of groupBy(pairs, keyFn).entries()) {
    out[k] = computeClusterStats(v);
  }
  return out;
}

/**
 * Slippage stats over fills.jsonl. Each fill carries `slippage_bps` (sim_executor
 * records it as (fill - limit)/limit for buys, (limit - fill)/limit for sells —
 * positive bps means we paid *up* vs our limit).
 */
function computeSlippageStats(fills, windowStart) {
  const rows = fills.filter(f => {
    if (f.slippage_bps == null) return false;
    if (!windowStart) return true;
    const t = new Date(f.filled_at || 0);
    return !isNaN(t.getTime()) && t >= windowStart;
  });
  if (!rows.length) return { count: 0 };
  const byAction = { BUY: [], SELL: [] };
  for (const f of rows) (byAction[f.action] ||= []).push(f.slippage_bps);
  const summarize = (arr) => {
    if (!arr.length) return null;
    const sorted = [...arr].sort((a, b) => a - b);
    return {
      count: arr.length,
      avg_bps: round(arr.reduce((a, b) => a + b, 0) / arr.length, 2),
      median_bps: round(sorted[Math.floor(sorted.length / 2)], 2),
      worst_bps: round(Math.max(...arr), 2),
    };
  };
  return {
    count: rows.length,
    overall: summarize(rows.map(f => f.slippage_bps)),
    buys: summarize(byAction.BUY || []),
    sells: summarize(byAction.SELL || []),
  };
}

function parseArgs(argv) {
  const args = { window: null, since: null, format: 'text' };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    const eq = (p) => a.startsWith(p + '=') ? a.slice(p.length + 1) : null;
    if (a === '--window') args.window = argv[++i];
    else if (eq('--window')) args.window = eq('--window');
    else if (a === '--since') args.since = argv[++i];
    else if (eq('--since')) args.since = eq('--since');
    else if (a === '--format') args.format = argv[++i];
    else if (eq('--format')) args.format = eq('--format');
    else if (a === '-h' || a === '--help') args.help = true;
    else throw new Error(`unknown argument: ${a}`);
  }
  if (!['text', 'json'].includes(args.format)) {
    throw new Error(`--format must be text|json, got ${args.format}`);
  }
  return args;
}

function padLeft(s, w) { s = String(s); return s.length >= w ? s : ' '.repeat(w - s.length) + s; }
function padRight(s, w) { s = String(s); return s.length >= w ? s : s + ' '.repeat(w - s.length); }

async function main() {
  let args;
  try { args = parseArgs(process.argv.slice(2)); }
  catch (e) { process.stderr.write(`error: ${e.message}\n`); return 2; }

  if (args.help) {
    process.stdout.write('Usage: node scripts/journal_stats.mjs [--window Nd|Nw] [--since YYYY-MM-DD] [--format=text|json]\n');
    return 0;
  }

  const store = await openStore();
  const [records, fills] = await Promise.all([store.listJournal(), store.listFills()]);
  if (!records.length) {
    process.stderr.write('no journal entries\n');
    return 2;
  }

  let windowStart;
  try { windowStart = parseWindow(args.window, args.since); }
  catch (e) { process.stderr.write(`error: ${e.message}\n`); return 2; }

  let pairs = pairTrades(records);

  if (windowStart) {
    pairs = pairs.filter(p => {
      const ts = p.exit.timestamp || '';
      const t = new Date(ts);
      if (isNaN(t.getTime())) return false;
      return t >= windowStart;
    });
  }

  if (!pairs.length) {
    process.stderr.write('no closed trades in window\n');
    return 1;
  }

  const overall = computeClusterStats(pairs);

  const groups = {
    by_trade_type: mapGroups(pairs, p => p.entry.trade_type || 'unknown'),
    by_theme_tag: mapGroups(pairs, p => p.entry.theme_tag || 'untagged'),
    by_sector: mapGroups(pairs, p => p.entry.sector || p.exit.sector || 'unknown'),
    by_exit_reason: mapGroups(pairs, p => p.exit.exit_reason || 'unknown'),
    by_conviction: mapGroups(pairs, p => convictionBucket(p.entry.conviction)),
    by_failure_mode: mapGroups(pairs, p =>
      `${p.exit.catalyst_occurred ?? 'unknown'}|${p.exit.mechanism_worked ?? 'unknown'}`
    ),
  };

  const slippage = computeSlippageStats(fills, windowStart);

  const payload = {
    generated_at: new Date().toISOString().replace(/\.\d+Z$/, 'Z'),
    window: args.window || args.since || 'all-time',
    n_closed: pairs.length,
    overall,
    groups,
    slippage,
  };

  if (args.format === 'json') {
    process.stdout.write(JSON.stringify(payload, null, 2) + '\n');
    return 0;
  }

  const lines = [];
  lines.push(`📊 JOURNAL STATS — window: ${payload.window}   closed trades: ${payload.n_closed}`);
  lines.push('='.repeat(72));
  lines.push(
    `Overall: win ${overall.win_rate_pct}%  avg ${overall.avg_pnl_pct}%  ` +
    `expectancy ${overall.expectancy_pct}%  avg days held ${overall.avg_days_held}`
  );
  for (const [groupName, rows] of Object.entries(groups)) {
    lines.push('');
    lines.push(`-- ${groupName} --`);
    const sorted = Object.entries(rows).sort((a, b) => b[1].count - a[1].count);
    for (const [key, r] of sorted) {
      lines.push(
        `  ${padRight(String(key), 40)} n=${padLeft(r.count, 3)}  ` +
        `win ${padLeft(r.win_rate_pct, 5)}%  ` +
        `avg ${padLeft(r.avg_pnl_pct.toFixed(2), 6)}%  ` +
        `exp ${padLeft(r.expectancy_pct.toFixed(2), 6)}%`
      );
    }
  }
  if (slippage.count) {
    lines.push('');
    lines.push('-- slippage (bps vs. limit; + = paid up) --');
    const fmt = (s) => s ? `n=${padLeft(s.count, 3)} avg ${padLeft(s.avg_bps, 7)}  med ${padLeft(s.median_bps, 7)}  worst ${padLeft(s.worst_bps, 7)}` : '(none)';
    lines.push(`  overall  ${fmt(slippage.overall)}`);
    lines.push(`  buys     ${fmt(slippage.buys)}`);
    lines.push(`  sells    ${fmt(slippage.sells)}`);
  }
  process.stdout.write(lines.join('\n') + '\n');
  return 0;
}

main().then(
  (code) => process.exit(code),
  (err) => { process.stderr.write(`FATAL: ${err.stack || err.message || err}\n`); process.exit(2); }
);
